src/codegen/mod.rs

```diff
@@ -0,0 +1,2000 @@
+        ForeignModBuilder {
+            inner: ast::ForeignMod {
+                abi: abi,
+                items: vec![],
+            }
+        }
+    }
+
+    fn with_foreign_item(mut self, item: ast::ForeignItem) -> Self {
+        self.inner.items.push(item);
+        self
+    }
+
+    #[allow(dead_code)]
+    fn with_foreign_items<I>(mut self, items: I) -> Self
+        where I: IntoIterator<Item=ast::ForeignItem>
+    {
+        self.inner.items.extend(items.into_iter());
+        self
+    }
+
+    fn build(self, ctx: &BindgenContext) -> P<ast::Item> {
+        use syntax::codemap::DUMMY_SP;
+        P(ast::Item {
+            ident: ctx.rust_ident(""),
+            id: ast::DUMMY_NODE_ID,
+            node: ast::ItemKind::ForeignMod(self.inner),
+            vis: ast::Visibility::Public,
+            attrs: vec![],
+            span: DUMMY_SP,
+        })
+    }
+}
+
+struct ArrayTyBuilder {
+    len: usize,
+}
+
+// TODO: Aster seem to lack a array builder?
+impl ArrayTyBuilder {
+    fn new() -> Self {
+        ArrayTyBuilder {
+            len: 0,
+        }
+    }
+
+    fn with_len(mut self, len: usize) -> Self {
+        self.len = len;
+        self
+    }
+
+    fn build(self, ty: P<ast::Ty>) -> P<ast::Ty> {
+        use syntax::codemap::DUMMY_SP;
+        let size =
+            ast::LitKind::Int(self.len as u64,
+                              ast::LitIntType::Unsigned(ast::UintTy::Us));
+        let size = ast::ExprKind::Lit(P(respan(DUMMY_SP, size)));
+        let array_kind = ast::TyKind::FixedLengthVec(ty,
+            P(ast::Expr {
+                id: ast::DUMMY_NODE_ID,
+                node: size,
+                span: DUMMY_SP,
+                attrs: ast::ThinVec::new(),
+            })
+        );
+
+        P(ast::Ty {
+            id: ast::DUMMY_NODE_ID,
+            node: array_kind,
+            span: DUMMY_SP,
+        })
+    }
+}
+
+struct BlobTyBuilder {
+    layout: Layout,
+}
+
+impl BlobTyBuilder {
+    fn new(layout: Layout) -> Self {
+        BlobTyBuilder {
+            layout: layout,
+        }
+    }
+
+    fn build(self) -> P<ast::Ty> {
+        use std::cmp;
+
+        let ty_name = match self.layout.align {
+            8 => "u64",
+            4 => "u32",
+            2 => "u16",
+            1 | _ => "u8",
+        };
+        let data_len = if ty_name == "u8" {
+            self.layout.size
+        } else {
+            self.layout.size / cmp::max(self.layout.align, 1)
+        };
+
+        let inner_ty = aster::AstBuilder::new().ty().path().id(ty_name).build();
+        if data_len == 1 {
+            inner_ty
+        } else {
+            ArrayTyBuilder::new().with_len(data_len).build(inner_ty)
+        }
+    }
+}
+
+/// A trait to convert a rust type into a pointer, optionally const, to the same
+/// type.
+///
+/// This is done due to aster's lack of pointer builder, I guess I should PR
+/// there.
+trait ToPtr {
+    fn to_ptr(self, is_const: bool, span: Span) -> P<ast::Ty>;
```

Comment:
Bindgen only tracks whether a field is const or not, so while I agree it would be a cleaner interface, I think changing all the callsites from:

```rust
ty.to_ptr(item.is_const(), ctx.span())
```

to 

```rust
ty.to_ptr(if item.is_const() { ast::Mutability::Mutable } else { ast::Mutability::Immutable }, ctx.span())
```

is counter-productive. I could add a method to the item that returns the actual mutability, but that means either introducing the `ir` module to `syntax::ast` code, or creating another artificial trait to implement it there.